

# Goal: This script analyzes Gastrodia tuber PE RNA-seq data generated by Lee lab
# Input: feacutreCounts read counts; including 3 top and 3 bottom (infected by armillaria gallica) 
# Output: DEG list
# Conclusion: 
# Ref: 
# Author: Chia-Yi Cheng, Jia-Yun Chong
# Last updated: 2022-07-02

# 0. Environment ---------------------------------------------------------------

setwd("C:/R_CJY/WGCNA_test")

library(EDASeq)
library(edgeR)
library(WGCNA)
library(RUVSeq)
library(RColorBrewer)

rm(list=ls())

# 1. Loading data --------------------------------------------------------------

df_ori <- read.table("GelataTuber.featureCounts.txt", header=T, row.name=1, check.names=F)
colnames(df_ori) # sample names
rownames(df_ori) # gene names

#1.1  Labeling -----------------------------------------------------------------

fc_condition <- factor(rep(c("up","basal"),each=3),levels=c("up","basal"))
fc_condition

fc_rep <- rep(c("1","2"),each=3)
fc_rep # why do we need rep, it has the same meaning as condition?

df_sampleInfo = data.frame(fc_condition,fc_rep)
df_sampleInfo

#1.2 Filter out lowly expressed genes -----

## convert input into a DGEList
dge_ori <- DGEList(counts=df_ori, group=fc_rep, remove.zeros = T) # removed 2889 rows
nrow(dge_ori) # 16604
head(dge_ori)
head(cpm(dge_ori))

## Filgering criteria: the average of 3 rep is above 2 CPM

#- use `cpm`(Counts per Million or Reads) function to normalize read counts by 
#- library size(the number of sequenced reads for a given sample), 
#- normalize_count = ori_count/(library size*10^{-6}).
#- After that, exchange row and column by `t`, 
#- now row are samples and column are genes.
#- `collapseRows` function select one representative row per group, in this case, 
#- our representative row is the average gene expression of each group(up 1 or 
#- bottom 2). 
list_collapse <- collapseRows(t(cpm(dge_ori)), rowGroup = fc_rep, 
                              rowID=colnames(dge_ori), method = "Average")
names(list_collapse) # "datETcollapsed" "group2row" "selectedRow"
head(t(list_collapse$selectedRow))
t(list_collapse$selectedRow)

#- we want to filter out gene which has low expression level in all sample, and 
#- we set a subjective value `lowest_cpm` to 2. If one of the group average is 
#- bigger than `lowest_cpm`, we will keep that gene.  
# ======
lowest_cpm = 2
# ======

mask_keep <- rowSums(t(list_collapse$datETcollapsed) >= lowest_cpm) >= 1
dge_filteredOutLowExprssion <- dge_ori[mask_keep, keep.lib.sizes=F]
nrow(dge_filteredOutLowExprssion) # 13775

## making sure there's no NA
table(is.na(dge_filteredOutLowExprssion)) # FALSE 2


# 2. Normalization -------------------------------------------------------------

#- i don't know why and how the following script normalized the data ...TAT...

## this is the universe (# of expressed genes) of this data set
list_genes <- rownames(dge_filteredOutLowExprssion)

#write.table(genes, quote=FALSE, row.names=F, col.names=F,sep = "\t",
#            file="PATH/FILENAME")

# Before normalization: convert filtered into a Expression Set
seqSet_ori <- newSeqExpressionSet(as.matrix(dge_filteredOutLowExprssion), 
                                  phenoData = data.frame(fc_condition, 
                                                         fc_rep, row.names = colnames(dge_filteredOutLowExprssion)))

# After normalization
#seqSet_n1 <- withinLaneNormalization(seqSet_ori) # withinLanNormalization needs
# feature data, which i don't know how to generate.

seqSet_normalized <- betweenLaneNormalization(seqSet_ori, which = "upper")

#RUVs: Estimating the factors of unwanted variation using replicate samples
differences<-makeGroups(fc_rep)
differences

seqSet_ <- RUVs(seqSet_normalized, list_genes, k=1, differences)
